### PR TITLE
Add ability to pass url state in `<CreateButton>`

### PR DESCRIPTION
**Pull request by mchaffotte**
Link to: https://github.com/marmelab/react-admin/pull/9303 due to too many modifications the PR should be based on next version.

- [x] add state props in button component
- [x] update e-commerce demo to show how to use it
- [x] update documentation

### CODE REVIEW

File: packages/ra-ui-materialui/src/button/CloneButton.tsx [hunk 1]
```diff
@@ -61,7 +61,7 @@ interface Props {
     scrollToTop?: boolean;
 }
 
-export type CloneButtonProps = Props & ButtonProps;
+export type CloneButtonProps = Props & ButtonProps<typeof Link>;
```
Comment (djhi) [inline packages/ra-ui-materialui/src/button/CloneButton.tsx:64]: You should exclude the `to` prop here or at least make it partial

File: examples/demo/src/reviews/ReviewCreate.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+const productOptionRenderer = (choice: any) => `${choice.reference}`;
+
+const ReviewCreate = () => {
+    const notify = useNotify();
+    const redirect = useRedirect();
+    const location = useLocation();
+
+    const onSuccess = (_: any) => {
+        const record = getRecordFromLocation(location);
```
Comment (djhi) [inline examples/demo/src/reviews/ReviewCreate.tsx:29]: You shouldn't have to do that as `onSuccess` is called with the result of `dataProvider.create`. If our dataProvider for this example does not return it, please fix it

**Review by djhi — CHANGES_REQUESTED** [review]
Almost there

File: docs/Create.md [hunk 3]
```diff
@@ -433,22 +433,15 @@ That means that if you want to create a link to a creation form, presetting *som
 {% raw %}
 ```jsx
 import * as React from 'react';
-import { Datagrid, useRecordContext } from 'react-admin';
-import { Button } from '@mui/material';
-import { Link } from 'react-router-dom';
+import { CreateButton, Datagrid, useRecordContext } from 'react-admin';
```
Comment (djhi) [inline docs/Create.md:436]: You should change the text of the previous paragraph at L431

File: examples/demo/src/reviews/ReviewCreate.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    getRecordFromLocation,
+} from 'react-admin';
+
+import StarRatingInput from './StarRatingInput';
+import { useLocation } from 'react-router';
+
+const customerOptionRenderer = (choice: any) =>
+    `${choice.first_name} ${choice.last_name}`;
+
+const productOptionRenderer = (choice: any) => `${choice.reference}`;
```
Comment (djhi) [inline examples/demo/src/reviews/ReviewCreate.tsx:21]:
```suggestion
```

File: examples/demo/src/reviews/ReviewCreate.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            <SimpleForm>
+                <ReferenceInput source="customer_id" reference="customers">
+                    <AutocompleteInput
+                        optionText={customerOptionRenderer}
+                        validate={required()}
+                    />
+                </ReferenceInput>
+                <ReferenceInput source="product_id" reference="products">
+                    <AutocompleteInput
+                        optionText={productOptionRenderer}
```
Comment (djhi) [inline examples/demo/src/reviews/ReviewCreate.tsx:49]:
```suggestion
                        optionText="reference"
```

You can provide a property name here for simple cases

File: examples/demo/src/reviews/ReviewCreate.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    useRedirect,
+    getRecordFromLocation,
+} from 'react-admin';
+
+import StarRatingInput from './StarRatingInput';
+import { useLocation } from 'react-router';
+
+const customerOptionRenderer = (choice: any) =>
+    `${choice.first_name} ${choice.last_name}`;
+
```
Comment (djhi) [inline examples/demo/src/reviews/ReviewCreate.tsx:20]: We usually put utilities like this at the end of the file. Important things first

File: examples/demo/src/reviews/StarRatingInput.tsx [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+const StyledRating = styled(Rating)({
+    '& .MuiRating-iconFilled': {
+        color: '#000',
+    },
+    '& .MuiRating-iconHover': {
+        color: '#000',
+    },
+});
+
```
Comment (djhi) [inline examples/demo/src/reviews/StarRatingInput.tsx:13]: Same, please move this at the end of the file

File: packages/ra-ui-materialui/src/button/CloneButton.tsx [hunk 8]
```diff
@@ -29,10 +29,10 @@ export const CloneButton = (props: CloneButtonProps) => {
                           search: stringify({
                               source: JSON.stringify(omitId(record)),
                           }),
-                          state: { _scrollToTop: scrollToTop },
                       }
                     : pathname
             }
+            state={{ _scrollToTop: scrollToTop }}
```
Comment (djhi) [inline packages/ra-ui-materialui/src/button/CloneButton.tsx:35]: This probably means we introduced a breaking change even though passing the previous `to` to the `Button` without a `component={Link}` didn't work before. I'm not sure what we should do here. @fzaninotto @slax57 @adguernier?

**Review by djhi — CHANGES_REQUESTED** [review]

File: examples/demo/src/reviews/ReviewCreate.tsx [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import { useLocation } from 'react-router';
+
+const ReviewCreate = () => {
+    const notify = useNotify();
+    const redirect = useRedirect();
+    const location = useLocation();
+
+    const onSuccess = (_: any) => {
+        const record = getRecordFromLocation(location);
+        notify(`Changes saved`);
```
Comment (fzaninotto) [inline examples/demo/src/reviews/ReviewCreate.tsx:25]: use an existing translation message instead - the EE demo is multilingual.

File: examples/demo/src/reviews/ReviewCreate.tsx [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+import StarRatingInput from './StarRatingInput';
+import { useLocation } from 'react-router';
+
+const ReviewCreate = () => {
+    const notify = useNotify();
+    const redirect = useRedirect();
+    const location = useLocation();
+
+    const onSuccess = (_: any) => {
+        const record = getRecordFromLocation(location);
```
Comment (fzaninotto) [inline examples/demo/src/reviews/ReviewCreate.tsx:24]: Why do you get the record from the location and not the record from the onSuccess callback parameter?

File: packages/ra-ui-materialui/src/button/CloneButton.tsx (hunk 8 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/button/CloneButton.tsx:35]: I agree that this is a breaking change and we should not do it. Instead, we must deprecate the support for the to prop in button, but keep the logic there.

File: packages/ra-ui-materialui/src/button/CreateButton.tsx [hunk 11]
```diff
@@ -41,12 +44,17 @@ const CreateButton = (props: CreateButtonProps) => {
     const isSmall = useMediaQuery((theme: Theme) =>
         theme.breakpoints.down('md')
     );
+    const state = merge(
+        scrollStates[String(scrollToTop)],
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/button/CreateButton.tsx:48]: security warning

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: examples/demo/src/reviews/ReviewCreate.tsx (hunk 10 above)
Comment (mchaffotte) [reply examples/demo/src/reviews/ReviewCreate.tsx:24]: I used the location in order to keep the possibility to create straightly reviews. If I use the parameter of the callback I do not know from where I come from and the redirection works in all cases.

Comment (djhi) [reply examples/demo/src/reviews/ReviewCreate.tsx:24]: In short: to handle cases where you don't come from a product edition view

Comment (fzaninotto) [reply examples/demo/src/reviews/ReviewCreate.tsx:24]: ah, I get it: the record isn't the created `review`, but a `product`. To clarify, you should perhaps name the variable `product`.

File: examples/demo/src/reviews/ReviewCreate.tsx [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+            redirect(`/reviews`);
+        }
+    };
+
+    return (
+        <Create mutationOptions={{ onSuccess }}>
+            <SimpleForm>
+                <ReferenceInput source="customer_id" reference="customers">
+                    <AutocompleteInput
+                        optionText={customerOptionRenderer}
```
Comment (fzaninotto) [inline examples/demo/src/reviews/ReviewCreate.tsx:38]: you should rely on the `recordRepresentation` instead.

File: packages/ra-ui-materialui/src/auth/AuthError.tsx [hunk 13]
```diff
@@ -19,7 +20,7 @@ export const AuthError = (props: AuthErrorProps) => {
             <div className={AuthErrorClasses.message}>
                 <h1>{translate(title, { _: title })}</h1>
                 <div>{translate(message, { _: message })}</div>
-                <Button to="/login" label="ra.auth.sign_in">
+                <Button component={Link} to="/login" label="ra.auth.sign_in">
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/auth/AuthError.tsx:23]: I believe this change is no longer necessary?

File: packages/ra-ui-materialui/src/button/Button.tsx [hunk 14]
```diff
@@ -54,8 +56,8 @@ export const Button = (props: ButtonProps) => {
                     className={className}
                     color={color}
                     size="large"
-                    {...rest}
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/button/Button.tsx:57]: Why did you switch the order?

File: packages/ra-ui-materialui/src/button/Button.tsx [hunk 15]
```diff
@@ -137,7 +141,9 @@ const StyledButton = styled(MuiButton, {
     },
 });
 
-const getLinkParams = (locationDescriptor?: LocationDescriptor | string) => {
+export const getLinkParams = (
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/button/Button.tsx:144]: Why do you need to export it? I don't see it used anywhere else.

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-ui-materialui/src/button/Button.tsx (hunk 14 above)
Comment (mchaffotte) [reply packages/ra-ui-materialui/src/button/Button.tsx:57]: In order to not overlap the value of 'state' prop coming from CreateButton (for example)

File: packages/ra-ui-materialui/src/auth/AuthError.tsx (hunk 13 above)
Comment (mchaffotte) [reply packages/ra-ui-materialui/src/auth/AuthError.tsx:23]: This writing is the new behaviour so it seems to be good to use it.

Comment (djhi) [reply packages/ra-ui-materialui/src/auth/AuthError.tsx:23]: I believe this didn't work before this PR anyway as the `Button` did not set the `component` prop automatically

File: packages/ra-ui-materialui/src/button/Button.tsx (hunk 15 above)
Comment (mchaffotte) [reply packages/ra-ui-materialui/src/button/Button.tsx:144]: Thanks I missed to add the behaviour in CreateButton

File: examples/demo/src/reviews/ReviewCreate.tsx [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+import StarRatingInput from './StarRatingInput';
+import { useLocation } from 'react-router';
+
+const ReviewCreate = () => {
+    const notify = useNotify();
+    const redirect = useRedirect();
+    const location = useLocation();
+
+    const onSuccess = (_: any) => {
+        const record = getRecordFromLocation(location);
```
Comment (djhi) [inline examples/demo/src/reviews/ReviewCreate.tsx:24]: It's not a product, it's an object containing default values for the new review. I would leave it as it is or rename it to `formInitialValues`

File: packages/ra-ui-materialui/src/button/Button.tsx (hunk 15 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/button/Button.tsx:144]: Can you remove the `export` now?

**Review by djhi — CHANGES_REQUESTED** [review]

File: examples/demo/src/reviews/ReviewCreate.tsx (hunk 16 above)
Comment (mchaffotte) [reply examples/demo/src/reviews/ReviewCreate.tsx:24]: If the parameter name is _,  it is due to a lint warning unused parameter. I prefer to let it like that. It is the review but from this parameter I have no clue to know if the review comes from a simple creation or a creation from a product.

Comment (djhi) [reply examples/demo/src/reviews/ReviewCreate.tsx:24]: @fzaninotto and I were referring to the `record` variable ;)

File: docs/Create.md [hunk 17]
```diff
@@ -463,29 +456,22 @@ export default PostList = () => (
...
-import { Link } from 'react-router-dom';
+import { CreateButton, useRecordContext } from 'react-admin';
 
 const CreateRelatedCommentButton = () => {
     const record = useRecordContext();
     return (
-        <Button
-            component={Link}
+        <CreateButton
+            resource="reviews"
```
Comment (slax57) [inline docs/Create.md:470]: Why reviews and not comments? :thinking: 

```suggestion
            resource="comments"
```

File: packages/ra-ui-materialui/src/button/CreateButton.tsx [hunk 18]
```diff
@@ -41,29 +47,34 @@
     const isSmall = useMediaQuery((theme: Theme) =>
         theme.breakpoints.down('md')
     );
+    const state = merge(scrollStates[String(scrollToTop)], initialState);
```
[the remark this answers is not in the document]
Comment (slax57) [reply packages/ra-ui-materialui/src/button/CreateButton.tsx:45]: You can probably replace the JS base object with a `Map` instead to mitigate this risk

**Review by slax57 — COMMENTED** [review]
Otherwise, LGTM. :muscle: 

Also: please rename the PR title to make it clearer this PR actually implements a new feature, and does not only update the docs (and the demo) :wink:

File: examples/demo/src/reviews/ReviewCreate.tsx [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+            redirect(`/reviews`);
+        }
+    };
+
+    return (
+        <Create mutationOptions={{ onSuccess }}>
+            <SimpleForm>
+                <ReferenceInput source="customer_id" reference="customers">
+                    <AutocompleteInput validate={required()} />
+                </ReferenceInput>
```
Comment (slax57) [inline examples/demo/src/reviews/ReviewCreate.tsx:38]: One more thing: the created reviews miss a status. You can probably use `defaultValues` to set the default status to be `'pending'`.

**Review by slax57 — CHANGES_REQUESTED** [review]

File: packages/ra-ui-materialui/src/button/CreateButton.tsx [hunk 20]
```diff
@@ -41,50 +42,56 @@ const CreateButton = (props: CreateButtonProps) => {
...
```
Comment (slax57) [inline packages/ra-ui-materialui/src/button/CreateButton.tsx:45]: The way to access values inside a `Map` is using `get()`

```suggestion
    const state = merge(scrollStates.get(String(scrollToTop)), initialState);
```

**Review by slax57 — CHANGES_REQUESTED** [review]

**Review by slax57 — APPROVED** [review]

**Pull request merged**